### ISSUE TITLE
[Discussion] Vector4, and other 4D math types

### ISSUE THREAD

**Issue by aaronfranke** [godotengine/godot#20907]
Potential use cases for a Vector4 class that I can think of:

* Parameters for noise functions, simplex etc: https://github.com/godotengine/godot/issues/1268

* For use with shaders: https://github.com/godotengine/godot/issues/9134 and https://github.com/godotengine/godot/issues/2855#issuecomment-406846872

* Representing an Axis-Angle 3D rotation (not the same as Quaternions, simpler, less capable)

* Represeting 3D vectors in Direction-Magnitude format (direction is normalized).

* Simply for allowing the user to store 4 values together with a built-in struct.

There's likely much more uses for Vector4 than I'm thinking of. I'm not desperate for it, but I would find a Vector4 class useful. What matters is what Godot users would need it to do. It would be engine bloat implementing Vector4 "just because". So I'm opening this discussion to ask around about if users would find Vector4 useful and what kind of things they would use it for.

If you can think of a use case for Vector4 that you'd use, feel free to post it here.

**Comment by aaronfranke** [comment]
@AndreaCatania Bullet has `btVector4`. Are there physics use cases for Vector4 in core?

**Comment by AndreaCatania** [comment]
In bullet I didn't used it, but I used it for the Flex Engine implementation

**Comment by groud** [comment]
I don't think this deserves to be in the core. The need for them is very limited, as even in the core we don't need them (we have the quaternion class for case 3)

**Comment by Illiander** [comment]
I don't really see why this isn't core.

Just for linking with the shader language vec4s if nothing else.

Unless Quats already do this, and can be passed to shaders, etc... as arbitrary 4-tuples.

**Comment by syskrank** [comment]
That would be really useful for many cases. 
Besides, Vector4 is a core type in many engines and (could be) used directly for producing more effects with shaders, storing complex data and scripting fancy math stuff. 

It won't hurt anyone or anything to have this implemented.

**Comment by Jummit** [comment]
> It won't hurt anyone or anything to have this implemented.

Reduz stated that submissions will be `refused by default` and that `bloat accumulates quickly`, and this doesn't seem like a feature that is needed ([follow the diagram here](https://godotengine.org/article/will-your-contribution-be-merged-heres-how-tell)).

**Comment by aaronfranke** [comment]
While the debate for the usefulness of Vector4 continues, if anyone is interested in having Vector4 in their own projects, I created [a C# math library](https://github.com/aaronfranke/GodotExtraMath) which includes Vector4 among other things. You still won't be able to use it as a parameter for Godot's built-in methods, at least not directly.
